src/terminal/adapter/SixelParser.cpp

```diff
@@ -0,0 +1,850 @@
+    _lastFlushTime = steady_clock::now();
+
+    // Prior to the VT340, the background was filled as soon as the sixel
+    // definition was started, because the initial raster attributes could
+    // not be altered.
+    if (_conformanceLevel < 3)
+    {
+        _fillImageBackground();
+    }
+}
+
+void SixelParser::_resizeImageBuffer(const til::CoordType requiredHeight)
+{
+    const auto requiredSize = (_imageCursor.y + requiredHeight) * _imageMaxWidth;
+    if (static_cast<size_t>(requiredSize) > _imageBuffer.size())
+    {
+        static constexpr auto transparentPixel = IndexedPixel{ .transparent = true };
+        _imageBuffer.resize(requiredSize, transparentPixel);
+    }
+}
+
+void SixelParser::_fillImageBackground()
+{
+    if (_backgroundFillRequired) [[unlikely]]
+    {
+        _backgroundFillRequired = false;
+
+        const auto backgroundHeight = std::min(_backgroundSize.height, _availablePixelHeight);
+        const auto backgroundWidth = std::min(_backgroundSize.width, _availablePixelWidth);
+        _resizeImageBuffer(backgroundHeight);
+
+        // When a background fill is requested, we prefill the buffer with the 0
+        // color index, up to the boundaries set by the raster attributes (or if
+        // none were given, up to the page boundaries). The actual image output
+        // isn't limited by the background dimensions though.
+        static constexpr auto backgroundPixel = IndexedPixel{};
+        const auto backgroundOffset = _imageCursor.y * _imageMaxWidth;
+        auto dst = std::next(_imageBuffer.begin(), backgroundOffset);
+        for (auto i = 0; i < backgroundHeight; i++)
+        {
+            std::fill_n(dst, backgroundWidth, backgroundPixel);
+            std::advance(dst, _imageMaxWidth);
+        }
+
+        _imageWidth = std::max(_imageWidth, backgroundWidth);
+    }
+}
+
+void SixelParser::_writeToImageBuffer(int sixelValue, int repeatCount)
+{
+    // On terminals that support the raster attributes command (which sets the
+    // background size), the background is only drawn when the first sixel value
+    // is received. So if we haven't filled it yet, we need to do so now.
+    _fillImageBackground();
+
+    // Then we need to render the 6 vertical pixels that are represented by the
+    // bits in the sixel value. Although note that each of these sixel pixels
+    // may cover more than one device pixel, depending on the aspect ratio.
+    const auto targetOffset = _imageCursor.y * _imageMaxWidth + _imageCursor.x;
+    auto imageBufferPtr = std::next(_imageBuffer.begin(), targetOffset);
+    repeatCount = std::min(repeatCount, _imageMaxWidth - _imageCursor.x);
+    for (auto i = 0; i < 6; i++)
+    {
+        if (sixelValue & 1)
+        {
+            auto repeatAspectRatio = _pixelAspectRatio;
+            do
+            {
+                std::fill_n(imageBufferPtr, repeatCount, _foregroundPixel);
+                std::advance(imageBufferPtr, _imageMaxWidth);

```

Comment:
Here, the same thing happens. Similarly, I'd move the `+ targetOffset` down here into the `fill_n` call.